pkg/services/meta/blocks.go

```diff
@@ -35,46 +38,78 @@ func (m *Meta) handleBlock(b *block.Header) error {
 	m.stM.RLock()
 	defer m.stM.RUnlock()
 
+	evsByStorage := make(map[*containerStorage][]objEvent)
 	for _, n := range res.Application {
 		ev, err := parseObjNotification(n)
 		if err != nil {
 			l.Error("invalid object notification received", zap.Error(err))
 			continue
 		}
 
-		s, ok := m.storages[ev.cID]
+		if magic := uint32(ev.network.Uint64()); magic != m.magicNumber {
+			l.Warn("skipping object notification with wrong magic number", zap.Uint32("expected", m.magicNumber), zap.Uint32("got", magic))
+		}
+
+		st, ok := m.storages[ev.cID]
 		if !ok {
 			l.Debug("skipping object notification", zap.Stringer("inactual container", ev.cID))
 			continue
 		}
 
-		err = m.handleObjectNotification(s, ev)
-		if err != nil {
-			l.Error("handling object notification", zap.Error(err))
-			continue
+		evsByStorage[st] = append(evsByStorage[st], ev)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		for st, ee := range evsByStorage {
+			st.putMPTIndexes(ee)
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		var internalWg errgroup.Group
+		internalWg.SetLimit(1024)
+
+		for st, evs := range evsByStorage {
+			internalWg.Go(func() error {
+				err := st.putRawIndexes(ctx, evs, m.net)
+				if err != nil {
+					l.Error("failed to put raw indexes", zap.String("storage", st.path), zap.Error(err))
+				}
+
+				// do not stop other routines ever
+				return nil
+			})
 		}
 
-		l.Debug("handled object notification successfully", zap.Stringer("cID", ev.cID), zap.Stringer("oID", ev.oID))
-	}
+		// errors are logged, no errors are returned to WG
+		_ = internalWg.Wait()
+	}()
 
-	for _, st := range m.storages {
+	wg.Wait()
+
+	for st := range evsByStorage {
 		// TODO: parallelize depending on what can parallelize well
 
 		st.m.Lock()
 
 		root := st.mpt.StateRoot()
 		st.mpt.Store.Put([]byte{rootKey}, root[:])
 		p := st.path
-		if st.opsBatch != nil {
-			_, err := st.mpt.PutBatch(mpt.MapToMPTBatch(st.opsBatch))
-			if err != nil {
-				st.m.Unlock()
-				return fmt.Errorf("put batch for %d block to %q storage: %w", ind, p, err)
-			}
 
-			st.opsBatch = nil
+		_, err := st.mpt.PutBatch(mpt.MapToMPTBatch(st.mptOpsBatch))
+		if err != nil {
+			st.m.Unlock()
+			return fmt.Errorf("put batch for %d block to %q storage: %w", ind, p, err)
 		}
 
+		maps.Clear(st.mptOpsBatch)
```

Comment:
But `clear()` is available for maps since Go 1.21.

Reply:
whoops, ok